supervision/detection/core.py

```diff
@@ -1028,18 +1030,26 @@ def with_nms(
 
         if class_agnostic:
             predictions = np.hstack((self.xyxy, self.confidence.reshape(-1, 1)))
-            indices = non_max_suppression(
-                predictions=predictions, iou_threshold=threshold
+        else:
+            assert self.class_id is not None, (
+                "Detections class_id must be given for NMS to be executed. If you"
+                " intended to perform class agnostic NMS set class_agnostic=True."
+            )
+            predictions = np.hstack(
+                (
+                    self.xyxy,
+                    self.confidence.reshape(-1, 1),
+                    self.class_id.reshape(-1, 1),
+                )
             )
-            return self[indices]
 
-        assert self.class_id is not None, (
-            "Detections class_id must be given for NMS to be executed. If you intended"
-            " to perform class agnostic NMS set class_agnostic=True."
-        )
+        if hasattr(self, "mask") and self.mask is not None:
```

Comment:
`Detections` always has an attribute `mask`, but it can be equal to `None` or `np.ndarray`. So we can simplify this condition to `self.mask is not None`.